.github/workflows/bump-cli.yml

```diff
@@ -0,0 +1,44 @@
+name: Bump CLI version
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+  schedule:
+  - cron: 0 0 */14 * * # run every 14 days
+
+jobs:
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 1
+          token: ${{ secrets.WORKFLOW_TOKEN }}
```

Comment:
And it looks like this is causing the workflow to fail since `WORKFLOW_TOKEN` is not defined.

Reply:
If we clone the repo without this token, the `push` step (in the create-pr action) gets rejected by the remote because GitHub refuses to accept any changes to `workflow` files unless we specifically provide a token that allows it.  For example [here](https://github.com/tjgurwara99/vscode-codeql/actions/runs/3699789862/jobs/6267519174) in my forks workflow, you can see in the logs

```
+ git push --force origin bump-cli
[67](https://github.com/tjgurwara99/vscode-codeql/actions/runs/3699789862/jobs/6267519174#step:4:70)
To https://github.com/tjgurwara99/vscode-codeql
[68](https://github.com/tjgurwara99/vscode-codeql/actions/runs/3699789862/jobs/6267519174#step:4:71)
 ! [remote rejected] bump-cli -> bump-cli (refusing to allow a GitHub App to create or update workflow `.github/workflows/main.yml` without `workflows` permission)
[69](https://github.com/tjgurwara99/vscode-codeql/actions/runs/3699789862/jobs/6267519174#step:4:72)
error: failed to push some refs to 'https://github.com/tjgurwara99/vscode-codeql'
[70](https://github.com/tjgurwara99/vscode-codeql/actions/runs/3699789862/jobs/6267519174#step:4:73)
Error: Process completed with exit code 1.
```

Reply:
Although this is technically possible by using one of our bot PATs, I would recommend against it. So far we have avoided storing any PATs in the secrets for public CodeQL repos, which avoids any risk of those tokens being compromised even if we accidentally include an overly permissive workflow that contributors can run on PRs.

Could you try either:
- Giving the `workflows` permission to the `GITHUB_TOKEN` (if that's possible)
- Avoiding changes to the workflow file itself in the automatically generated PR, instead modifying a separate file that records the version numbers and gets parsed by the first job in the generated workflow to determine the matrix (we have some prior art on this if that's useful)

Reply:
I think we could also use a different approach:
- We can use a GitHub App for creating the commit and the PR. This is not trivial since it requires using the GraphQL API to create commits, but I can create an action for doing this.
- Store the required secrets (app ID, installation ID and private key) in environment secrets. We can restrict the environment to only allow running it on the `main` branch, so it will not be possible for contributors to run the workflow without their change being in `main`. By using a GitHub app, we are also able to lock down which repositories it can access and which scopes it has. This is also possible using fine-grained access tokens, but as an admin on this repo you probably still have more access than the GitHub app.

I've already implemented something like this in the past, so I'll try to extract that into a GitHub Action which we can use here to better show what I have in mind.

Reply:
@tjgurwara99 I've created a PR on your fork for using a GitHub App: https://github.com/tjgurwara99/vscode-codeql/pull/10. It uses an action on my personal account (https://github.com/koesie10/create-app-pr-action), so it might be that we need to follow the release process for open source releases if we want to have this in the GitHub organization.

We would probably also need to follow a process for creating a GitHub Actions app. I've placed more information about what exactly we would need in the PR.

Reply:
> Giving the workflows permission to the GITHUB_TOKEN (if that's possible)

That isn't possible according to the documentation I think. Let me know if I missed something in the docs though 😅 - I have a habit of skim reading docs and jumping to conclusions. 

> Avoiding changes to the workflow file itself in the automatically generated PR, instead modifying a separate file that records the version numbers and gets parsed by the first job in the generated workflow to determine the matrix (we have some prior art on this if that's useful)

I would personally prefer that actually. The way I see it, if we don't need to modify anything in the workflow file, then the default `GH_TOKEN` suffices for us. @adityasharad you mentioned there is prior art to doing using external data in actions, could you point me to it? I would much prefer that approach 😄 

> @tjgurwara99 I've created a PR on your fork for using a GitHub App: https://github.com/tjgurwara99/vscode-codeql/pull/10. It uses an action on my personal account (https://github.com/koesie10/create-app-pr-action), so it might be that we need to follow the release process for open source releases if we want to have this in the GitHub organization.

Thanks @koesie10, I have left some comments on the PR - though I'm reluctant to use an app because, well there are more things that we own if we create it. I would much rather do this cleanly in place than create an extra repository that we'd have to own and most likely forget about. Though again, if external actions are allowed then I'm more than open to it. Though it still needs some credentials and I'd argue that they are equally as risky as PAT exposure (though maybe less so in theory). So I'm still in favour of not having any new secrets - I'll try to concoct something later today and re-review your custom action (I reviewed it briefly but need to give it a thorough review 😄). Let me know what are people's thoughts on having either btw.